pkg/controller/nodeipam/ipam/cidrset/cidr_set.go

```diff
@@ -263,3 +275,161 @@ func (s *CidrSet) getIndexForIP(ip net.IP) (int, error) {
 
 	return 0, fmt.Errorf("invalid IP: %v", ip)
 }
+
+// Map manages a set of CIDR ranges from which blocks of IPs can
+// be allocated from.
+type Map struct {
+	sync.Mutex
+	clusterCIDR   *net.IPNet
+	nodeMask      net.IPMask
+	maxCIDRs      uint64
+	nextCandidate uint64
+	used          map[uint64]bool
+}
+
+var _ Interface = &Map{}
+
+// NewMapCIDRSet creates a new CidrSet based on a map[uint64]bool.
+func NewMapCIDRSet(clusterCIDR *net.IPNet, subNetMaskSize int) (*Map, error) {
+	clusterMask := clusterCIDR.Mask
+	clusterMaskSize, clusterMaskLen := clusterMask.Size()
+	// Normalize the cluster IP
+	clusterCIDR.IP.Mask(clusterCIDR.Mask)
+	// Find the maximum number
+	var maxCIDRs uint64
+	if (clusterCIDR.IP.To4() == nil) && (subNetMaskSize-clusterMaskSize >= clusterMapSubnetMaxDiff) {
+		return nil, ErrCIDRSetSubNetTooBig
+	}
+	maxCIDRs = 1 << uint64(subNetMaskSize-clusterMaskSize)
+	return &Map{
+		clusterCIDR: clusterCIDR,
+		maxCIDRs:    maxCIDRs,
+		nodeMask:    net.CIDRMask(subNetMaskSize, clusterMaskLen),
+		used:        make(map[uint64]bool),
+	}, nil
+}
+
+// Occupy marks the given CIDR range as used.
+func (m *Map) Occupy(cidr *net.IPNet) error {
+	m.Lock()
+	defer m.Unlock()
+	// validate that CIDR belongs to the Cluster CIDR ranges
+	if err := m.validate(cidr); err != nil {
+		return err
+	}
+	// obtain the IP key
+	ipKey := m.calculateIPOffset(cidr.IP)
+
+	if _, ok := m.used[ipKey]; ok {
```

Comment:
it may pass a cidr with a mask higher than the node-mask-cidr, ie.

cidr = 192.168.0.0/26 and node-mask-cidr 30, it has to occupy 192.168.0.0 - 192.168.0.64